File: sdk-rust/src/schema_check.rs
//! Schema-version check executed by `GatedhouseFactory::create`.

use crate::database::Database;
use crate::error::GatedhouseError;

pub(crate) const EXPECTED_VERSION: i32 = 1;

pub(crate) fn verify(database: &dyn Database) -> Result<(), GatedhouseError> {
    let mut conn = database.connection()?;
    let row = conn.query_opt(
        "SELECT 1 FROM information_schema.tables \
         WHERE table_schema = 'gatedhouse' AND table_name = 'schema_versions'",
        &[],
    )?;
    if row.is_none() {
        return Err(GatedhouseError::SchemaNotInitialized);
    }
    let row = conn.query_one(
        "SELECT COALESCE(MAX(version), 0) FROM gatedhouse.schema_versions",
        &[],
    )?;
    let current: i32 = row.get(0);
    if current < EXPECTED_VERSION {
        return Err(GatedhouseError::SchemaOutOfDate {
            current_version: current,
            expected_version: EXPECTED_VERSION,
        });
    }
    Ok(())
}

File: sdk-python/gatedhouse/_schema_check.py
# Copyright (c) 2026 12vectors.com
# SPDX-License-Identifier: MIT
# See the LICENSE file in the repository root for the full license text.

"""Schema-version check executed by ``GatedhouseFactory.create``."""

from __future__ import annotations

from psycopg import errors as pg_errors

from ._database import Database
from ._exceptions import SchemaNotInitializedError, SchemaOutOfDateError

EXPECTED_VERSION = 1


def verify(database: Database) -> None:
    try:
        with database.connection() as conn, conn.cursor() as cur:
            cur.execute(
                "SELECT 1 FROM information_schema.tables "
                "WHERE table_schema = 'gatedhouse' "
                "  AND table_name = 'schema_versions'"
            )
            if cur.fetchone() is None:
                raise SchemaNotInitializedError()
            cur.execute(
                "SELECT COALESCE(MAX(version), 0) "
                "FROM gatedhouse.schema_versions"
            )
            row = cur.fetchone()
            current = row[0] if row else 0
            if current < EXPECTED_VERSION:
                raise SchemaOutOfDateError(current, EXPECTED_VERSION)
    except pg_errors.Error as e:
        raise SchemaNotInitializedError() from e
